swift/apple/FirezoneKit/Sources/FirezoneKit/Views/MenuBar.swift

```diff
@@ -236,176 +164,164 @@ public final class MenuBar: NSObject, ObservableObject {
 
-  @objc private func signOutButtonTapped() {
-    do { try store.signOut() } catch { Log.error(error) }
-  }
+    store.$status
+      .receive(on: DispatchQueue.main)
+      .sink(receiveValue: { [weak self] _ in
+        guard let self = self else { return }
+        self.handleStatusChanged()
+      }).store(in: &cancellables)
 
-  @objc private func grantPermissionMenuItemTapped() {
-    Task {
-      do {
-        // If we get here, it means either system extension got disabled or
-        // our VPN configuration got removed. Since we don't know which, reinstall
-        // the system extension here too just in case. It's a no-op if already
-        // installed.
-        try await store.installSystemExtension()
-        try await store.grantVPNPermission()
-      } catch {
-        Log.error(error)
-        await macOSAlert.show(for: error)
-      }
-    }
+    updateChecker.$updateAvailable
+      .receive(on: DispatchQueue.main)
+      .sink(receiveValue: { [weak self] _ in
+        guard  let self = self else { return }
+        self.handleUpdateAvailableChanged()
+      }).store(in: &cancellables)
   }
 
-  @objc private func settingsButtonTapped() {
-    AppView.WindowDefinition.settings.openWindow()
+  func handleFavoritesChanged() {
+    // When the user clicks to add or remove a favorite, the menu will close anyway, so just recreate the whole
+    // menu. This avoids complex logic when changing in and out of the "nothing is favorited" special case.
+    populateResourceMenus([])
+    populateResourceMenus(store.resourceList.asArray())
+    updateResourcesMenuItems()
   }
 
-  @objc private func adminPortalButtonTapped() {
-    guard let url = URL(string: store.settings.authBaseURL)
-    else { return }
-
-    Task { await NSWorkspace.shared.openAsync(url) }
+  func handleResourceListChanged() {
+    populateResourceMenus(store.resourceList.asArray())
+    updateResourcesMenuItems()
   }
 
-  @objc private func updateAvailableButtonTapped() {
-    Task { await NSWorkspace.shared.openAsync(UpdateChecker.downloadURL()) }
+  func handleStatusChanged() {
+    updateStatusItemIcon()
+    updateSignInMenuItems()
+    quitMenuItem.title = {
+      switch store.status {
+      case .connected, .connecting:
+        return "Disconnect and Quit"
+      default:
+        return "Quit"
+      }
+    }()
   }
 
-  @objc private func documentationButtonTapped() {
-    let url = URL(string: "https://www.firezone.dev/kb?utm_source=macos-client")!
-
-    Task { await NSWorkspace.shared.openAsync(url) }
+  func handleUpdateAvailableChanged() {
+    updateStatusItemIcon()
+    refreshUpdateItem()
   }
 
-  @objc private func supportButtonTapped() {
-    let url = URL(string: "https://www.firezone.dev/support?utm_source=macos-client")!
-
-    Task { await NSWorkspace.shared.openAsync(url) }
-  }
+  func populateResourceMenus(_ newResources: [Resource]) {
+    // If we have no favorites, then everything is a favorite
```

Comment:
This logic should be in `Store` IMO.

Reply:
A lot of this logic will probably be cleaned even further when #7771 is implemented. Then, we can just use one `SessionView` where on iOS it's the current thing and on macOS it's a MenuBarExtra view.

Still, it might be nice to have a computed var in Store that reflects the grouping of Internet, favorited, and non-favorited Resources.

Reply:
I'll save that for a future PR. Hoping to get this out soon to fix a UI update bug (and possibly a mem leak) on iOS.